enarx-keep-sgx-shim/src/handler.rs

```diff
@@ -22,37 +27,90 @@ extern "C" {
     ) -> u64;
 }
 
+pub trait Print<T: ?Sized> {
+    fn print(&mut self, data: &T);
+}
+
```

Comment:
Could Handler just implement `core::fmt::Write`? Then we get the print macros.

Reply:
https://www.youtube.com/watch?v=YJDNbhrGFy0

Reply:
I attempted to implement this yesterday with some success, but encountered two problems.

1. While we get the use of macros like `writeln!()`, this offloads error handling to every call site (the compiler warns us that `Result` is unused). However, we don't really have a good way to handle this error should it arise. We could potentially work around this with a local macro (maybe named `eprintln!()`?) that calls `attacked()` if an error occurs.

2. Some forms of format arguments don't work and it is unclear why. Frustratingly, debug formatting doesn't work (`{:?}`) and causes `writeln!()` to fail after outputting all previous data.

Unless we can solve (2) quickly, I suggest we make this a debt issue.

Reply:
@connorkuehl I made a `debug!()` wrapper for handling. So (1) is solvable. However, I also discovered that attempting to format strings causes a page fault, for a reason I don't understand.

Are you okay with making this a `debt` issue?

Reply:
@npmccallum Yeah, let's make this a `debt` issue.

The debug formatter is a head scratcher. Theoretically format_args! is a compile-time primitive that is non-allocating, and I always just thought that the output of a debug format specifier is a `str` anyway, so I can't imagine why it's fussing.

Reply:
https://github.com/enarx/enarx/issues/469

Reply:
Here is my println!() ...https://github.com/enarx/enarx/blob/demo2020/enarx-keep-sev-shim/src/arch/x86_64/serial/syscall_serial.rs

Reply:
needs cleanup.. historically grown...